### DISCUSSION TITLE
创建session修改参数报错

### DISCUSSION THREAD

**Discussion by dakeleblack** [alibaba/GraphScope#344] [Q&A] [answered]
Graphscope默认参数创建session时出现内存不够的异常，修改参数后报如下错，貌似不支持参数修改？

graphscope.set_option(k8s_coordinator_mem="2Gi")
s=graphscope.session()

Initializing graphscope session with parameters: {'addr': None, 'cluster_type': 'k8s', 'num_workers': 2, 'preemptive': True, 'k8s_namespace': None, 'k8s_service_type': 'NodePort', 'k8s_gs_image': 'registry.cn-hongkong.aliyuncs.com/graphscope/graphscope:0.4.0', 'k8s_etcd_image': 'quay.io/coreos/etcd:v3.4.13', 'k8s_image_pull_policy': 'IfNotPresent', 'k8s_image_pull_secrets': [], 'k8s_gie_graph_manager_image': 'registry.cn-hongkong.aliyuncs.com/graphscope/maxgraph_standalone_manager:0.4.0', 'k8s_zookeeper_image': 'zookeeper:3.4.14', 'k8s_coordinator_cpu': 1.0, 'k8s_coordinator_mem': '2Gi', 'k8s_etcd_num_pods': 3, 'k8s_etcd_cpu': 0.5, 'k8s_etcd_mem': '128Mi', 'k8s_zookeeper_cpu': 0.5, 'k8s_zookeeper_mem': '256Mi', 'k8s_gie_graph_manager_cpu': 1.0, 'k8s_gie_graph_manager_mem': '4Gi', 'k8s_vineyard_daemonset': '', 'k8s_vineyard_cpu': 0.5, 'k8s_vineyard_mem': '512Mi', 'vineyard_shared_mem': '4Gi', 'k8s_engine_cpu': 0.5, 'k8s_engine_mem': '4Gi', 'k8s_mars_worker_cpu': 0.5, 'k8s_mars_worker_mem': '4Gi', 'k8s_mars_scheduler_cpu': 0.5, 'k8s_mars_scheduler_mem': '2Gi', 'with_mars': False, 'k8s_volumes': {}, 'k8s_waiting_for_delete': False, 'timeout_seconds': 600, 'dangling_timeout_seconds': 600, 'k8s_client_config': {}}
2021-05-24 10:59:15,535 [INFO][cluster:446]: Launching coordinator...
---------------------------------------------------------------------------
ApiException                              Traceback (most recent call last)
/opt/conda/lib/python3.8/site-packages/graphscope/deploy/kubernetes/cluster.py in start(self)
    659 
--> 660             self._create_services()
    661             time.sleep(1)

/opt/conda/lib/python3.8/site-packages/graphscope/deploy/kubernetes/cluster.py in _create_services(self)
    583     def _create_services(self):
--> 584         self._create_coordinator()
    585 

/opt/conda/lib/python3.8/site-packages/graphscope/deploy/kubernetes/cluster.py in _create_coordinator(self)
    489         targets.append(
--> 490             self._app_api.create_namespaced_deployment(
    491                 self._namespace, coordinator_builder.build()

/opt/conda/lib/python3.8/site-packages/kubernetes/client/api/apps_v1_api.py in create_namespaced_deployment(self, namespace, body, **kwargs)
    341         kwargs['_return_http_data_only'] = True
--> 342         return self.create_namespaced_deployment_with_http_info(namespace, body, **kwargs)  # noqa: E501
    343 

/opt/conda/lib/python3.8/site-packages/kubernetes/client/api/apps_v1_api.py in create_namespaced_deployment_with_http_info(self, namespace, body, **kwargs)
    436 
--> 437         return self.api_client.call_api(
    438             '/apis/apps/v1/namespaces/{namespace}/deployments', 'POST',

/opt/conda/lib/python3.8/site-packages/kubernetes/client/api_client.py in call_api(self, resource_path, method, path_params, query_params, header_params, body, post_params, files, response_type, auth_settings, async_req, _return_http_data_only, collection_formats, _preload_content, _request_timeout, _host)
    347         if not async_req:
--> 348             return self.__call_api(resource_path, method,
    349                                    path_params, query_params, header_params,

/opt/conda/lib/python3.8/site-packages/kubernetes/client/api_client.py in __call_api(self, resource_path, method, path_params, query_params, header_params, body, post_params, files, response_type, auth_settings, _return_http_data_only, collection_formats, _preload_content, _request_timeout, _host)
    179         # perform request and return response
--> 180         response_data = self.request(
    181             method, url, query_params=query_params, headers=header_params,

/opt/conda/lib/python3.8/site-packages/kubernetes/client/api_client.py in request(self, method, url, query_params, headers, post_params, body, _preload_content, _request_timeout)
    390         elif method == "POST":
--> 391             return self.rest_client.POST(url,
    392                                          query_params=query_params,

/opt/conda/lib/python3.8/site-packages/kubernetes/client/rest.py in POST(self, url, headers, query_params, post_params, body, _preload_content, _request_timeout)
    273              body=None, _preload_content=True, _request_timeout=None):
--> 274         return self.request("POST", url,
    275                             headers=headers,

/opt/conda/lib/python3.8/site-packages/kubernetes/client/rest.py in request(self, method, url, query_params, headers, body, post_params, _preload_content, _request_timeout)
    232         if not 200 <= r.status <= 299:
--> 233             raise ApiException(http_resp=r)
    234 

ApiException: (422)
Reason: Unprocessable Entity
HTTP response headers: HTTPHeaderDict({'Audit-Id': 'c17a08d3-9fb5-4377-ba5f-dec41471fa5a', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Date': 'Mon, 24 May 2021 10:59:16 GMT', 'Content-Length': '540'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Deployment.apps \"coordinator-zgixua\" is invalid: spec.template.spec.containers[0].resources.requests: Invalid value: \"4Gi\": must be less than or equal to memory limit","reason":"Invalid","details":{"name":"coordinator-zgixua","group":"apps","kind":"Deployment","causes":[{"reason":"FieldValueInvalid","message":"Invalid value: \"4Gi\": must be less than or equal to memory limit","field":"spec.template.spec.containers[0].resources.requests"}]},"code":422}

The above exception was the direct cause of the following exception:

K8sError                                  Traceback (most recent call last)
<ipython-input-4-68c078aae0c8> in <module>
----> 1 s=graphscope.session()

/opt/conda/lib/python3.8/site-packages/graphscope/client/utils.py in wrapper(*args, **kwargs)
    154             func.__defaults__ = tuple(new_defaults)
    155 
--> 156             return_value = func(*args, **kwargs)
    157 
    158             # Restore original defaults.

/opt/conda/lib/python3.8/site-packages/graphscope/client/session.py in __init__(self, config, cluster_type, addr, num_workers, preemptive, k8s_namespace, k8s_service_type, k8s_gs_image, k8s_etcd_image, k8s_gie_graph_manager_image, k8s_zookeeper_image, k8s_image_pull_policy, k8s_image_pull_secrets, k8s_coordinator_cpu, k8s_coordinator_mem, k8s_etcd_num_pods, k8s_etcd_cpu, k8s_etcd_mem, k8s_zookeeper_cpu, k8s_zookeeper_mem, k8s_gie_graph_manager_cpu, k8s_gie_graph_manager_mem, k8s_vineyard_daemonset, k8s_vineyard_cpu, k8s_vineyard_mem, vineyard_shared_mem, k8s_engine_cpu, k8s_engine_mem, k8s_mars_worker_cpu, k8s_mars_worker_mem, k8s_mars_scheduler_cpu, k8s_mars_scheduler_mem, k8s_volumes, k8s_waiting_for_delete, timeout_seconds, dangling_timeout_seconds, with_mars, **kw)
    489         # create and connect session
    490         with CaptureKeyboardInterrupt(self.close):
--> 491             self._connect()
    492 
    493         self._disconnected = False

/opt/conda/lib/python3.8/site-packages/graphscope/client/session.py in _connect(self)
    826         # launching graphscope service
    827         if self._launcher is not None:
--> 828             self._launcher.start()
    829             self._coordinator_endpoint = self._launcher.coordinator_endpoint
    830 

/opt/conda/lib/python3.8/site-packages/graphscope/deploy/kubernetes/cluster.py in start(self)
    667             self._dump_coordinator_failed_status()
    668             self.stop()
--> 669             raise K8sError(
    670                 "Error when launching Coordinator on kubernetes cluster"
    671             ) from e

K8sError: 'Error when launching Coordinator on kubernetes cluster'

**Comment by lidongze0629** [comment, marked as answer]
Hi, @bitao666, thank you for posting this question in the discussions, and welcome to the GraphScope discussions.

From your log, the exception `resources.requests: Invalid value: "4Gi": must be less than or equal to memory limit'. ` occurs when resource limits less than requests. 

In your current version of GraphScope, the coordinator memory requests defaults to '4Gi'.  So your parameters `k8s_coordinator_mem="2Gi"` cannot pass the API check.  You can solve this problem by passing `preemptive=False` together.

```python
graphscope.set_option(k8s_coordinator_mem="2Gi", preemptive=False)
s=graphscope.session()
```

**[From GraphScope Doc](https://graphscope.io/docs/reference/session.html)**
> preemptive (bool, optional) – If True, GraphScope instance will treat resource params (e.g. k8s_coordinator_cpu) as limits and provide the minimum available value as requests, but this will make pod has a Burstable QOS, which can be preempted by other pods with high QOS. Otherwise, it will set both requests and limits with the same value.

**Reply by dakeleblack** [reply]
it works, thanks for your answer
